test/client_test.go

```diff
@@ -124,3 +169,52 @@ func TestNewWithDiscovery(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, 3, controller.getCount)
 }
+
+func TestCircuitBreaker(t *testing.T) {
+	controller := ClientCircuitTest{}
+	s := server.New(&controller, "")
+	testServer := httptest.NewServer(s.Handler)
+	defer testServer.Close()
+	hystrix.Flush()
+	c := client.New(testServer.URL)
+	c.SetCircuitBreakingDebug(false)
+
+	// the circuit should open after 20 consecutive failed attempts (this is
+	// the default volume threshold, after which health can open the circuit)
+	controller.down = true
+	var connAttempts int64
+	ctx := httptrace.WithClientTrace(context.Background(),
+		&httptrace.ClientTrace{
+			GetConn: func(hostPort string) {
+				atomic.AddInt64(&connAttempts, 1)
+			},
+		})
+	for count := 0; count < 100; count++ {
+		_, err := c.CreateBook(ctx, &models.Book{})
+		assert.Error(t, err)
+	}
+	assert.Equal(t, int64(20), connAttempts)
+
+	// we should see an attempt go through after five seconds
+	circuitOpened := time.Now()
+	for _ = range time.Tick(100 * time.Millisecond) {
```

Comment:
this can infinite loop if there's a bug. should we add some catch to prevent that